File: apps/worker/src/contract/messages.ts
import type { AnalysisAttemptId, CountsBasis, UnitSystem } from '@gbd/db';
import * as v from 'valibot';
import { CHART_KEY_PATTERN } from './layout.ts';
import { CHILD_FAILURE_REASONS, COUNTS_BASES, UNIT_SYSTEMS } from './names.ts';

export class ContractError extends Error {}

const MONTH_PATTERN = /^\d{4}-(0[1-9]|1[0-2])$/;
const SHA_256_PATTERN = /^[0-9a-f]{64}$/;
const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
// `ai_cost_usd` is `numeric(10,4)`; float64 can't round-trip it, so it crosses as a string.
const COST_USD_PATTERN = /^\d{1,6}\.\d{4}$/;

const nonEmptyString = v.pipe(v.string(), v.nonEmpty());
const wholeNumber = v.pipe(v.number(), v.integer(), v.minValue(0));

const opaqueObject = v.custom<Record<string, unknown>>(
  (value) => typeof value === 'object' && value !== null && !Array.isArray(value),
  'expected a JSON object',
);

const RunManifestSchema = v.strictObject({
  analysisAttemptId: v.pipe(v.string(), v.regex(UUID_PATTERN)),
  report: v.strictObject({
    name: v.nullable(nonEmptyString),
    siteName: v.nullable(nonEmptyString),
    countsBasis: v.picklist(COUNTS_BASES),
    unitSystem: v.picklist(UNIT_SYSTEMS),
    monthlyCounts: v.pipe(
      v.record(v.pipe(v.string(), v.regex(MONTH_PATTERN)), wholeNumber),
      v.check((counts) => Object.keys(counts).length > 0, 'needs at least one month'),
    ),
  }),
  inputFile: v.strictObject({
    originalFilename: nonEmptyString,
    byteSize: v.pipe(v.number(), v.integer(), v.minValue(1)),
    checksumSha256: v.pipe(v.string(), v.regex(SHA_256_PATTERN)),
  }),
});

const ProgressSchema = v.strictObject({
  sequence: v.pipe(v.number(), v.integer(), v.minValue(1)),
});

const ChildResultSchema = v.strictObject({
  analysisAttemptId: v.pipe(v.string(), v.regex(UUID_PATTERN)),
  // Chart keys, not paths — `layout.ts` derives the filename.
  charts: v.pipe(
    v.array(v.pipe(v.string(), v.regex(CHART_KEY_PATTERN))),
    v.checkItems(
      (chartKey, index, charts) => charts.indexOf(chartKey) === index,
      'chart keys must be unique',
    ),
  ),
  ai: v.strictObject({
    model: nonEmptyString,
    inputTokens: wholeNumber,
    outputTokens: wholeNumber,
    costUsd: v.pipe(v.string(), v.regex(COST_USD_PATTERN)),
    metadata: opaqueObject,
  }),
  resultMetadata: opaqueObject,
});

const ChildFailureSchema = v.strictObject({
  reason: v.picklist(CHILD_FAILURE_REASONS),
  detail: nonEmptyString,
  traceback: v.nullable(nonEmptyString),
});

export type RunManifest = v.InferOutput<typeof RunManifestSchema>;
export type Progress = v.InferOutput<typeof ProgressSchema>;
export type ChildResult = v.InferOutput<typeof ChildResultSchema>;
export type ChildFailure = v.InferOutput<typeof ChildFailureSchema>;

// `monthlyCounts` is `unknown` because Kysely hands back a `jsonb` column untyped.
export type RunManifestInput = {
  analysisAttemptId: AnalysisAttemptId;
  report: {
    name: string | null;
    siteName: string | null;
    countsBasis: CountsBasis;
    unitSystem: UnitSystem;
    monthlyCounts: unknown;
  };
  inputFile: { originalFilename: string; byteSize: number; checksumSha256: string };
};

export function buildRunManifest(input: RunManifestInput): RunManifest {
  return validate(RunManifestSchema, 'run.json', input);
}

// The real child parses `run.json` in Python; this exists only so this stack's own tests can
// round-trip what `buildRunManifest` produced and check it against the golden fixtures.
export function parseRunManifest(text: string): RunManifest {
  return parseDocument(RunManifestSchema, 'run.json', text);
}

export function parseProgress(text: string): Progress {
  return parseDocument(ProgressSchema, 'progress.json', text);
}

export function parseResult(text: string): ChildResult {
  return parseDocument(ChildResultSchema, 'result.json', text);
}

export function parseFailure(text: string): ChildFailure {
  return parseDocument(ChildFailureSchema, 'failure.json', text);
}

function parseDocument<TSchema extends v.GenericSchema>(
  schema: TSchema,
  documentName: string,
  text: string,
): v.InferOutput<TSchema> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (cause) {
    throw new ContractError(`${documentName}: not valid JSON`, { cause });
  }
  return validate(schema, documentName, parsed);
}

function validate<TSchema extends v.GenericSchema>(
  schema: TSchema,
  documentName: string,
  value: unknown,
): v.InferOutput<TSchema> {
  const result = v.safeParse(schema, value);
  if (result.success) return result.output;

  const problems = result.issues
    .map((issue) => `${v.getDotPath(issue) ?? '<root>'}: ${issue.message}`)
    .join('; ');
  throw new ContractError(`${documentName}: ${problems}`);
}

File: python/worker_child/src/worker_child/messages.py
"""The four documents that cross the run directory, and the child's half of them.

The child reads one and writes three, so this module is asymmetric on purpose: `run.json` gets a
parser, and the rest get payload builders. The parent's half is
`apps/worker/src/contract/messages.ts`.

Keys are camelCase across the seam, because the parent's fields come off Kysely rows already
camelCase and go straight back into camelCase columns. This module is where that translation
happens, once.
"""

import re
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from decimal import Decimal
from typing import Any

from worker_child.contract import (
    CHART_KEY_PATTERN,
    CHILD_FAILURE_REASONS,
    CONTRACT_VERSION,
    COUNTS_BASES,
    UNIT_SYSTEMS,
    ChildFailureReason,
    CountsBasis,
    UnitSystem,
)
from worker_child.parse import ContractError, root_fields

MONTH_PATTERN = re.compile(r"\d{4}-(0[1-9]|1[0-2])")
SHA_256_PATTERN = re.compile(r"[0-9a-f]{64}")

# Lowercase hex, matching what Postgres emits. Deliberately spelled out rather than deferring to
# each language's idea of a UUID, so both sides accept exactly the same set.
UUID_PATTERN = re.compile(r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}")


@dataclass(frozen=True, slots=True)
class ReportInputs:
    name: str | None
    site_name: str | None
    counts_basis: CountsBasis
    unit_system: UnitSystem
    monthly_counts: Mapping[str, int]


@dataclass(frozen=True, slots=True)
class InputFileFacts:
    original_filename: str
    byte_size: int
    checksum_sha256: str


@dataclass(frozen=True, slots=True)
class RunManifest:
    analysis_attempt_id: str
    report: ReportInputs
    input_file: InputFileFacts


@dataclass(frozen=True, slots=True)
class AiUsage:
    model: str
    input_tokens: int
    output_tokens: int
    # A Decimal, because `ai_cost_usd` is `numeric(10,4)` and it crosses the seam as a string.
    # A float would hand the parent 2.4712999999999997 for 2.4713.
    cost_usd: Decimal
    metadata: Mapping[str, Any]


def parse_run_manifest(text: str) -> RunManifest:
    """Read `input/run.json`, the one document the parent writes and the child reads."""
    root = root_fields("run.json", text, CONTRACT_VERSION)
    analysis_attempt_id = root.matching("analysisAttemptId", UUID_PATTERN)
    report = root.nested("report")
    input_file = root.nested("inputFile")

    manifest = RunManifest(
        analysis_attempt_id=analysis_attempt_id,
        report=ReportInputs(
            name=report.nullable_string("name"),
            site_name=report.nullable_string("siteName"),
            counts_basis=report.literal("countsBasis", COUNTS_BASES),
            unit_system=report.literal("unitSystem", UNIT_SYSTEMS),
            monthly_counts=report.counts("monthlyCounts", MONTH_PATTERN),
        ),
        input_file=InputFileFacts(
            original_filename=input_file.string("originalFilename"),
            byte_size=input_file.integer("byteSize", minimum=1),
            checksum_sha256=input_file.matching("checksumSha256", SHA_256_PATTERN),
        ),
    )

    report.done()
    input_file.done()
    root.done()
    return manifest


def progress_payload(sequence: int) -> dict[str, Any]:
    """Liveness, and nothing else.

    The parent measures a hang from this counter rather than from a timestamp or an mtime: a
    child rewriting identical content cannot look alive, and no clock crosses the seam.
    """
    if sequence < 1:
        raise ContractError(f"progress.json: sequence must be >= 1, got {sequence}")
    return {"contractVersion": CONTRACT_VERSION, "sequence": sequence}


def result_payload(
    *,
    analysis_attempt_id: str,
    charts: Sequence[str],
    ai: AiUsage,
    result_metadata: Mapping[str, Any],
) -> dict[str, Any]:
    """The success verdict.

    The PDF and the XLSX are not listed: they are mandatory by type, so a success that lacks
    either cannot be expressed. Charts are listed by key, and the parent derives each filename.
    """
    _require(UUID_PATTERN.fullmatch(analysis_attempt_id), "analysisAttemptId is not a uuid")
    for chart_key in charts:
        _require(
            CHART_KEY_PATTERN.fullmatch(chart_key),
            f"chart key '{chart_key}' is not snake_case",
        )
    _require(len(set(charts)) == len(charts), "chart keys must be unique")
    _require(ai.input_tokens >= 0 and ai.output_tokens >= 0, "token counts must not be negative")
    _require(ai.cost_usd >= 0, "cost must not be negative")

    return {
        "contractVersion": CONTRACT_VERSION,
        "analysisAttemptId": analysis_attempt_id,
        "charts": list(charts),
        "ai": {
            "model": ai.model,
            "inputTokens": ai.input_tokens,
            "outputTokens": ai.output_tokens,
            "costUsd": f"{ai.cost_usd:.4f}",
            "metadata": _opaque(ai.metadata, "ai.metadata"),
        },
        "resultMetadata": _opaque(result_metadata, "resultMetadata"),
    }


def failure_payload(
    *,
    reason: ChildFailureReason,
    detail: str,
    traceback: str | None = None,
) -> dict[str, Any]:
    """The failure verdict.

    No `analysisAttemptId`, unlike the success verdict: a child that cannot parse `run.json` does
    not know the attempt id, and reporting exactly that is what `contract_violation` is for. The
    parent knows which attempt it spawned from the directory it built.
    """
    _require(reason in CHILD_FAILURE_REASONS, f"'{reason}' is not a reason a child may claim")
    _require(bool(detail), "detail must not be empty")
    return {
        "contractVersion": CONTRACT_VERSION,
        "reason": reason,
        "detail": detail,
        "traceback": traceback,
    }


def _opaque(value: Mapping[str, Any], name: str) -> dict[str, Any]:
    """A bag the parent stores without inspecting, straight into a jsonb column.

    Anything the parent branches on is a declared field; anything else lands here. What actually
    goes in is unknown until the analysis library is ported — see `REQUIREMENTS.md`.
    """
    if not isinstance(value, Mapping):
        raise ContractError(f"{name} must be a JSON object")
    return dict(value)


def _require(condition: object, problem: str) -> None:
    if not condition:
        raise ContractError(problem)
